libportal-qt/filechooser.h

```diff
@@ -0,0 +1,140 @@
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * SPDX-License-Identifier: LGPL-3.0-only
+ */
+
+#ifndef LIBPORTALQT_FILECHOOSER_H
+#define LIBPORTALQT_FILECHOOSER_H
+
+#include "libportalqt_export.h"
+
+#include <QFlags>
+#include <QMap>
+#include <QStringList>
+#include <QSharedPointer>
+
+namespace Xdp
+{
+
+enum class OpenFileFlag
+{
+    None = 0x0,
+    Multiple = 0x1
+};
+Q_DECLARE_FLAGS(OpenFileFlags, OpenFileFlag)
+
+enum class SaveFileFlag
+{
+    None = 0x0
+};
+Q_DECLARE_FLAGS(SaveFileFlags, SaveFileFlag)
+
+class FileChooserFilterRulePrivate;
+
+class LIBPORTALQT_EXPORT FileChooserFilterRule
+{
+public:
+    enum class Type {
+        Pattern = 0,
+        Mimetype = 1
+    };
+
+    explicit FileChooserFilterRule();
+    FileChooserFilterRule(Type type, const QString &rule);
+    FileChooserFilterRule(const FileChooserFilterRule &other);
+    ~FileChooserFilterRule();
+
+    void operator=(const FileChooserFilterRule &other);
+
+    bool isValid() const;
+
+    Type type() const;
+    void setType(Type type);
+
+    QString rule() const;
+    void setRule(const QString &rule);
+private:
+    Q_DECLARE_PRIVATE(FileChooserFilterRule)
+
+    const QScopedPointer<FileChooserFilterRulePrivate> d_ptr;
+};
+typedef QList<FileChooserFilterRule> FileChooserFilterRules;
+
+class FileChooserFilterPrivate;
+
+class LIBPORTALQT_EXPORT FileChooserFilter
+{
+public:
+    explicit FileChooserFilter();
+    FileChooserFilter(const QString &label, const FileChooserFilterRules &rules);
+    FileChooserFilter(const FileChooserFilter &other);
+    ~FileChooserFilter();
+
+    void operator=(const FileChooserFilter &other);
+
+    bool isValid() const;
+
+    QString label() const;
+    void setLabel(const QString &label);
+
+    FileChooserFilterRules rules() const;
+    void addRule(const FileChooserFilterRule &rule);
+private:
+    Q_DECLARE_PRIVATE(FileChooserFilter)
+
+    const QScopedPointer<FileChooserFilterPrivate> d_ptr;
+};
+typedef QList<FileChooserFilter> FileChooserFilterList;
+
+class FileChooserChoicePrivate;
+
+class LIBPORTALQT_EXPORT FileChooserChoice
+{
+public:
+    explicit FileChooserChoice();
+    FileChooserChoice(const QString &id, const QString &label, const QMap<QString, QString> &options, const QString &selected = QString());
+    FileChooserChoice(const FileChooserChoice &other);
+    ~FileChooserChoice();
+
+    void operator=(const FileChooserChoice &other);
+
+    bool isValid() const;
+
+    QString id() const;
+    void setId(const QString &id);
+
+    QString label() const;
+    void setLabel(const QString &label);
+
+    QMap<QString, QString> options() const;
+    void addOption(const QString &id, const QString &label);
```

Comment:
Are these options really any string? Would it make sense to make them an enum or actual properties? or the idea is that they will get extended over time?

Reply:
Options are content of comboboxes:
```
        <varlistentry>
          <term>choices a(ssa(ss)s)</term>
          <listitem>
            <para>
              List of serialized combo boxes to add to the file chooser.
            </para>
            <para>
              For each element, the first string is an ID that will be returned
              with the response, the second string is a user-visible label. The
              a(ss) is the list of choices, each being an ID and a
              user-visible label. The final string is the initial selection,
              or "", to let the portal decide which choice will be initially selected.
              None of the strings, except for the initial selection, should be empty.
            </para>
            <para>
              As a special case, passing an empty array for the list of choices
              indicates a boolean choice that is typically displayed as a check
              button, using "true" and "false" as the choices.
            </para>
            <para>
              Example: [('encoding', 'Encoding', [('utf8', 'Unicode (UTF-8)'), ('latin15', 'Western')], 'latin15'), ('reencode', 'Reencode', [], 'false')]
            </para>
          </listitem>
        </varlistentry>
      </variablelist>
```

Reply:
ok :+1: